### DISCUSSION TITLE
Add support for hardware / external software

### DISCUSSION THREAD

**Discussion by holblin** [File-New-Project/EarTrumpet#747] [Ideas]
Hi,

I search to have a hardware volume mixer like one of theses: https://github.com/omriharel/deej/blob/master/community.md#mozza-v2

I found multiple solutions available but they all missed something. To cite some of them:
- Pc-Panel: The software is very limited. +analog input
- Turn Up: Still Kickstarter + unknown software + analog input
- NK2Tray: Analog input + some hardware switch unmap the setting
- GoXLR: Pricy + spaceship on a desk
- Deej: analog input or no support of bidirectional input

So, I could continue to explain why I don't like the existing solution during hours but the main piece of software I search is inside EarTrumpet. It's simple, provide all the functionality that I want (if you add the focused app volume) and just missing a hardware device to make it perfect :-)

To have an idea of what I would like to create: https://gfycat.com/wellinformedultimatehoverfly
This means, having the possibility to connect some hardware (through USB/serial?) that have access to:
- The current volume setting per "channel"
- The current sound played on "channel"
- The possibility to change the volume setting for a "channel"

Each channel would be a choice between:
- an output
- an audio group
- an audio
- the active audio app
- the active output 

On top of that, I would love to have physical buttons that could:
- Mute / Unmute / Toggle mute output
- Mute / Unmute / Toggle mute channel
- Mute / Unmute / Toggle mute active app
- Change the default output
- Change the output for an audio group

The cherry on the top would be to have the icon of the group transferred to the hardware to do some stuff like https://github.com/ho0ber/NK2Tray/tree/master/NK2Tray_Display

The idea is to be able to use motorized linear slides and or Rotary Encoder LED Ring which requires bidirectional communication (that deej doesn't have) to be aware of changes.

So, before receiving a big NO, my idea is not to push all of that into EarTrumpet but just to connect EarTrumpet to something that could handle the hardware part. I don't know if this could be done into an extension or if that needs some other program or if that's definitely something that needs to be done into a fork.

I would appreciate any help/guidance. I'm a developer even if I never did any: windows / c# development in the past.

**Comment by WolfganP** [comment]
Maybe adding an standard and widely use interface like MIDI may help interfacing EarTrumpet & external hardware or software bidirectionally, without increasing the support for every different piece of HW/SW around.

ie https://github.com/corrados/jamulus/ a low latency jam sessions via intenet enabler, allows the mixing channels volume and solo/grouping be controlled externally via MIDI, so it's up to the user to configure their own hardware/software to get the MIDI routing properly.

**Comment by holblin** [comment]
That's sound an amazing idea! I read a little bit about midi and this sounds to be a great solution. This will allow a full range of hardware! I only see one drawback, I'm pretty sure we couldn't get the icon and transfer them through midi 😓 but that's not a big deal since a text could be transferred (name of the program).

It would use a MIDI output and MIDI input to send and receive updates. According to that page, I could already have an idea of the commands used (more readable that the midi specification for a quick overview): https://www.noterepeat.com/articles/how-to/213-midi-basics-common-terms-explained

This would for example use a Continue Controller/Control Change/(CC) to set or get the level of the slider/knobs. This also could use a Continue Controller/Control Change/(CC) for the current volume sound playing (UV meter). For all the mute/unmute, this could use ON/OFF Messages - momentary vs toggle. And for switching between N items (like default speaker, audio output group, and so on), this could also use Continue Controller/Control Change/(CC). We could also transfer the name of the program with a text meta midi message (ref: https://www.recordingblogs.com/wiki/midi-text-meta-message).

I like a lot that idea, and I think that's a good way to do it. I'm wondering now if that should be developed into EarTrumpet with all the MIDI mapping or if that should be into a separate software that would communicate with EarTrumpet. WDYT?

**Comment by svenwml** [comment]
Currently, @skief and me are working on a solution for this. It's already working, but some more features and refactoring are required: https://www.youtube.com/watch?v=RihkBpO1qhU

**Comment by holblin** [comment]
Amazing! Could I help? At least for testing?

**Comment by skief** [comment]
> Amazing! Could I help? At least for testing?

You can check out the current state at https://github.com/svenwml/EarTrumpet/tree/midi-support and test it. 
Using CC Messages from sliders and rotary potentiometers should work but push buttons and rotary encoders do not work yet.
The main problem we have right now is that the code is still a bit messy and we are using some not so pretty workarounds to get the UI running.
